File: java/MergeSort/MergeSort.java
public class MergeSort {
    private static void mergeSort(int[] unsorted) {
        System.out.print("\n\nBefore: ");
        printArray(unsorted);
        mergeSort(unsorted, 0, unsorted.length);
        System.out.print("\nAfter: ");
        printArray(unsorted);
    }

    private static void mergeSort(int[] array, int startIndex, int endIndex) {
        //recursion base case is when startIndex and endIndex are equal
        if ((endIndex-startIndex) <= 1) return;

        //split array sections to sort in to halves
        int middle = ((endIndex - startIndex) / 2) + startIndex;
        mergeSort(array, startIndex, middle);
        mergeSort(array, middle, endIndex);

        //merge the respective halves back together
        mergeHalves(array, startIndex, middle, endIndex);
    }

    private static void mergeHalves(int[] array, int startIndex, int middle, int endIndex) {
        int[] temp = new int[array.length];
        int tempIndex = startIndex;

        int leftIndex = startIndex;
        int rightIndex = middle;

        while (leftIndex < middle && rightIndex < endIndex) {
            if (array[leftIndex] < array[rightIndex]) {
                temp[tempIndex++] = array[leftIndex++];
            } else {
                temp[tempIndex++] = array[rightIndex++];
            }
        }

        //any left over elements from either side can now be copied over to temp
        while (leftIndex < middle) temp[tempIndex++] = array[leftIndex++];
        while (rightIndex < endIndex) temp[tempIndex++] = array[rightIndex++];

        //now copy over the elements from temp back to array
        for (int i=startIndex;i<endIndex;i++)
            array[i] = temp[i];
    }

    private static void printArray(int[] array) {
        System.out.print("[");
        for (int i=0;i<array.length-1;i++) {
            System.out.print(array[i]+", ");
        }
        System.out.print(array[array.length-1]+"]");
    }

    public static void main(String[] argv) {
        int[] array1 = {3};
        mergeSort(array1);

        int[] array2 = {3,0};
        mergeSort(array2);

        int[] array3 = {0,-10,-1};
        mergeSort(array3);

        int[] array4 = {-1, -10, 5, 30034, 4, 35, 10, -300, 4};
        mergeSort(array4);

        int[] array5 = {0, 99, 42, 71, 1, 6, 9, 50, -4, 43, 100041};
        mergeSort(array5);
    }
}

File: python/merge_sort/merge_sort.py
import math

def merge_sort(my_array, start_index, end_index):

    # Handle recurse base case
    if ((end_index - start_index) <= 1):
        return

    middle = math.floor(((end_index - start_index) / 2) + start_index)
    #print("my_array[{}:{}], middle: {}".format(start_index, end_index, middle))
    merge_sort(my_array, start_index, middle)
    merge_sort(my_array, middle, end_index)

    #merge halves
    merge_halves(my_array, start_index, middle, end_index)


def merge_halves(my_array, start_index, middle, end_index):
    temp_array = [None] * len(my_array)
    temp_index = start_index
    left_index = start_index
    right_index = middle

    while ((left_index < middle) and (right_index < end_index)):
        if (my_array[left_index] < my_array[right_index]):
            temp_array[temp_index] = my_array[left_index]
            left_index += 1
        else:
            temp_array[temp_index] = my_array[right_index]
            right_index += 1
        temp_index += 1

    # now copy any outstanding elements
    while (left_index < middle):
        temp_array[temp_index] = my_array[left_index]
        temp_index += 1
        left_index += 1

    while (right_index < end_index):
        temp_array[temp_index] = my_array[right_index]
        temp_index += 1
        right_index += 1

    # Now copy the temp array values in to the real one
    temp_index = start_index
    while (temp_index < end_index):
        my_array[temp_index] = temp_array[temp_index]
        temp_index += 1

def begin_merge_sort(my_array):
    print("Before: {}".format(my_array.__str__()))
    merge_sort(my_array, 0, len(my_array))
    print("After: {}".format(my_array.__str__()))

def run_merge_sort():
    my_array = [3, 5, 0, -1, 45, 89]
    begin_merge_sort(my_array)
    
    my_array = [7, -1241231, -3 -342, 1134, 9, 0, 35, -90, 467, 23, 11, 0, 10, 345677654]
    begin_merge_sort(my_array)

if __name__ == "__main__":
    run_merge_sort()